Use scatter for check boxes and set facecolors correctly in check boxes and radio buttons

chahak13 opened the pull request:
ref: #24471 

With the current implementation, the boxes get stretched into rectangles if the aspect ratio is not maintained. To overcome this, the boxes are now created using scatter instead to maintain their shapes.

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

chahak13 commented on lib/matplotlib/widgets.py:
```diff
@@ -1072,9 +1071,18 @@ def set_active(self, index):
+                    p.get_facecolor(), colors.to_rgba("none"))
+                                else "none")
 
         if self.drawon:
             self.ax.figure.canvas.draw()
```
@anntzer I have a preliminary implementation for square check boxes but I'm running into one small issue. When I try running this, everything works as intended when we change the figure size (the check boxes remain square), but the unit test fails. For some reason, calling `set_active(0)` in code doesn't change the status of the checkboxes. The issue is arising in these lines. The status changes correctly based on the code that I added but after executing `self.ax.figure.canvas.draw()`, the colors for all the scatter markers change back to what they were initially set to be, and not the changed ones. This is happening only in the unit tests and not while actually checking it using a GUI backend. Any ideas/suggestions?

chahak13 replied:
I feel like I'm missing something but can't see it right now.

anntzer replied:
If you do some print-debugging and grep for places that could touch facecolors, you'll see that Collection.update_scalarmappable gets called at some point which causes the facecolors to be reset, but you'll need to figure out by yourself why (because I don't actually know immediately).

chahak13 replied:
That helps, thanks! I'll try figuring out why this is happening.

chahak13 replied:
@anntzer this is actually true for the radio buttons too! Small code to reproduce (based on the checkbox example)

```python
radio = RadioButtons(rax, labels, active=1, activecolor="k")
print(radio.value_selected)
print(radio._buttons.get_facecolor())
print("Changing active button from 1 to 0")
radio.set_active(0)
print(radio.value_selected)
print(radio._buttons.get_facecolor())
```
Output:
```
4 Hz
[[0. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 0. 0. 0.]]
Changing active button from 1 to 0
2 Hz
[[0. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 0. 0. 0.]]
```

This shows that the facecolors for the buttons don't change even after `set_active(0)` is called, even though `value_selected` is correct. The problem arises because of these lines

https://github.com/matplotlib/matplotlib/blob/a9c96200f7c3eac1923310086df96072813bac7d/lib/matplotlib/collections.py#L842-L845

Since it sets the flags such that it returns True the first time, 

https://github.com/matplotlib/matplotlib/blob/a9c96200f7c3eac1923310086df96072813bac7d/lib/matplotlib/collections.py#L894-L900

set the original colors again. Since `update_scalarmappable` gets called the first time they're drawn, when creating a figure, all functionalities work as everything works as expected after the second call. Even in this case, if there are any `set_active` calls after the first one, they work fine.

anntzer replied:
Thanks for the investigation, I have opened https://github.com/matplotlib/matplotlib/issues/24479 to track this.  Sorry for sending you down this rabbit hole, perhaps this was not so easy after all...

chahak13 replied:
Thanks, it was fun. But now with the question on hand, this bug stops me to run `get_status` for the check boxes correctly. I guess one workaround could be that I maintain an active array similar to `value_active` in radio buttons but that'll be a hack and I'm on the fence about having an extra array to handle that when it can be done with the colors itself. That'll pass the tests though, and everything works in a figure anyway. Suggestions?

oscargus commented on lib/matplotlib/widgets.py:
```diff
@@ -1002,43 +1002,24 @@ def __init__(self, ax, labels, actives=None):
-            p = Rectangle(xy=(x, y), width=w, height=h, edgecolor='black',
-                          facecolor=axcolor, transform=ax.transAxes)
+        ys = np.linspace(1, 0, len(labels)+2)[1:-1]
+        text_size = mpl.rcParams["font.size"] / 2
```
This seems unused? (Otherwise I would have asked why it is halved?)

chahak13 replied:
That's a mistake. It's supposed to be used for marker size. As to why it is halved, it's because radio buttons are also half the text size, that's why. I think it's arbitrary.

tacaswell commented:
I do not think that going with scatter is the correct approach here.  Looking at the internals of `Rectangle` we have a (private) knob to account for the aspect ratio https://github.com/matplotlib/matplotlib/blob/a9c96200f7c3eac1923310086df96072813bac7d/lib/matplotlib/patches.py#L720-L726

I suspect that making sure this gets set on `draw` is a much simpler way to solve this problem.

chahak13 commented:
@tacaswell  Just making sure that I'm understanding correctly, but if we do that then we're kinda working on the current implementation itself, right? In that case, we'll also have to make a call on should the checkbox size increase, while maintaining the square shape or not?

timhoffm commented:
> I do not think that going with scatter is the correct approach here.

@tacaswell why? Using markers instead of cobbling the box together from basic artists has some appeal. You don't have to bother with aspect. You get pixel snapping. You have only a single artist to change for updating the check state. And you can switch to more complex markers like a check symbol relatively easy.

ksunden commented on lib/matplotlib/widgets.py:
```diff
@@ -1072,9 +1071,18 @@ def set_active(self, index):
-        l1, l2 = self.lines[index]
-        l1.set_visible(not l1.get_visible())
-        l2.set_visible(not l2.get_visible())
+        if colors.same_color(
+                self._crosses.get_facecolor()[index], colors.to_rgba("none")
+        ):
+            self._crosses.get_facecolor()[index] = colors.to_rgba("k")
+        else:
+            self._crosses.get_facecolor()[index] = colors.to_rgba("none")
```
The point made in #24479 is that editing the output of the getter function is not guaranteed to _keep_ those changes (it works, sometimes, because it is not a copy that is returned, but there are other mechanisms at play)

more properly this would be:

```python
facecolors = self._crosses.get_faceolor()
facecolors[index] = colors.to_rgba("k") # or "none"
self._crosses.set_facecolor(facecolors)
```

This calls the setter which updates the internal state properly

(This also does need to be corrected in the radio button implementation)

anntzer commented:
From a quick look, _aspect_ratio_correction, introduced in #20839 for the rectangle selector widget, seems actually brittle, as 1) it will basically never work for polar plots (not a concern here, but one for rectangle selectors in screen space), and 2) it makes one of the two axises the "main" one relative to the other (I think for circles the radius will effectively be the radius in relative units along the x axis, not along the y axis), whereas the proper thing to do would be to have a radius in pixels (basically proportional to the font size, to make the radio button close to the size of the label).
If we really want to do this with individual artists we'd need Circle to support having different transforms for center and radius (a bit like how PathCollections have both a transform and an offset_transform, which is what we're effectively relying on here).

tacaswell commented:
Fair enough.  I missed that we did this for radio buttons and I am biased towards the minimal changes to solve the problem.   The consensus is clearly in favor of this so I take back my comment.

anntzer reviewed (changes requested):
I would rather not change the test image, see how I ensured that it still uses the old radiobuttons.  We can always reconsider and change both at the same time once the deprecation elapses, but the better way to test the new version is probably to use check_figures_equal like I did in test_radio_buttons.

anntzer commented on lib/matplotlib/widgets.py:
```diff
@@ -1100,6 +1113,24 @@ def disconnect(self, cid):
         self._observers.disconnect(cid)
 
+    @property
+    def rectangles(self):
```
This property needs to be deprecated, and you also need to handle `.lines`.

chahak13 replied:
Yes. Updated now.

chahak13 commented:
> I would rather not change the test image, see how I ensured that it still uses the old radiobuttons. We can always reconsider and change both at the same time once the deprecation elapses, but the better way to test the new version is probably to use check_figures_equal like I did in test_radio_buttons.

Ah gotcha. I'll take a look at your PR and do that. Thanks.

chahak13 commented:
The failing check is unhappy because I changed an image and reverted the test image for `test_check_radio_buttons`. Is there a recommended way to resolve that? Also, I'm not sure why the docs are failing, I did not make any changes there.

ksunden commented:
Docs test is because we fail on warnings and you got this warning:

https://app.circleci.com/pipelines/github/matplotlib/matplotlib/20207/workflows/95614b42-e1b0-4ead-850e-66805085a944/jobs/72159?invite=true#step-113-1153

which is a poor interaction with an old changelog entry which didn't use the full name of a reference.

https://github.com/matplotlib/matplotlib/blame/v3.6.2/doc/api/prev_api_changes/api_changes_2.1.0.rst#L425

This line should probably be updated to simply be ``lines`` (no reference, but codestyled) as it is in a heading (and the proper reference link is right below it anyway...)

anywho, because you are defining a function called "lines" and that reference was not fully specified, it had worked because there was only one target for "lines" previously, but now there is more than one.

ksunden commented:
As for the PR cleanliness check, are you familiar with `git rebase -i`? 

here is an article that goes over the basics, probably better than I could in an off hand github comment (though admittedly I only skimmed this article to make sure it wasn't outright incorrect):

https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

Essentially rather than committing, changing and then reverting that image in separate commits, squashing out the changes to the image entirely will make it so that there is not an extra version of the image in the canonical git history.

chahak13 commented:
Ah okay. That makes sense. Though, should I do anything in this PR to resolve that? That's not close to anything I'm working on here.

chahak13 commented:
Yes, I'm familiar with rebase, just did not want to go rebase stuff without knowing if that was fine since some people prefer not to do that. I'll rebase the commits to remove the image issue. Thanks!

ksunden commented:
since its a one line change that while it is unrelated to what you were directly working on, it is elevated to a warning by what you were doing, I'd just go ahead and fix it here to keep CI happy rather than opening a separate PR

chahak13 commented on doc/api/prev_api_changes/api_changes_2.1.0.rst:
```diff
@@ -422,7 +422,7 @@ The ``shading`` kwarg to `~matplotlib.axes.Axes.pcolor` has been
 
-Functions removed from the `.lines` module
+Functions removed from the :mod:`matplotlib.lines` module
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Ah okay. My bad

timhoffm commented on lib/matplotlib/tests/test_widgets.py:
```diff
@@ -1006,8 +1006,12 @@ def test_check_radio_buttons_image():
+    with pytest.warns(DeprecationWarning):
+        # Trigger old-style Rectangle check boxes
+        cb.rectangles
+        cb.lines
```
This line is never reached. You need to put in in a separate `warns()` context.

chahak13 replied:
The `cb.lines` line?

timhoffm replied:
Yes.

chahak13 replied:
Ah okay, thanks. I realized a funny thing then. Accessing rectangles is all that's required since that's all that changes from the older reference image with the rectangles. The line crosses and the scatter crosses are the exact same. Anyway, I've updated it, thanks!

QuLogic replied:
I think you'd need to create a new `CheckButtons` object to check that the `cb.lines` is falling back correctly.

chahak13 replied:
@QuLogic sorry, I was unavailable over the last week. This is to just check the fallback functionality or to also compare it with the reference image? Because if it is the former, then adding another test to just check the fallback would be better I believe.

chahak13 replied:
Okay, I changed things up a little, let me know if this makes sense. Earlier, only the property that was accessed (say `rectangles`) was being set, but if someone is working with that property, they would also expect the other one (here `lines`) to be set in a similar way. So now, whenever the property is accessed for the first time, it calls the other to set it. This way, there's consistency in the method used for drawing the boxes. As a consequence, this doesn't need an extra instance for checking fallback for `lines` as that gets checked by `rectangles` itself. 

If we want a test for the fallbacks, I can add 2 extra tests that just check those things.

QuLogic replied:
We should still check both, I think. Also, does accessing the other object trigger a second warning unnecessarily?

QuLogic replied:
You can wrap it in a `_api.suppress_matplotlib_deprecation_warning` context, or put the code in a separate private function that's called from both.

chahak13 replied:
@QuLogic I've updated and added 2 tests for checking the fallbacks. Let me know if this makes sense now!

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -1047,11 +1029,30 @@ def __init__(self, ax, labels, actives=None):
-                self.set_active(i)
-                break
+        pclicked = self.ax.transAxes.inverted().transform((event.x, event.y))
+        _, square_inds = self._squares.contains(event)
+        coords = self._squares.get_offset_transform().transform(
```
These two can go in the `else`, I think?

chahak13 commented on lib/matplotlib/widgets.py:
```diff
@@ -1002,43 +1002,25 @@ def __init__(self, ax, labels, actives=None):
+            c=["none" for _ in range(len(ys))],
+            linewidth=1, transform=ax.transAxes, edgecolor="k"
+        )
+        mask = [not x for x in actives]
```
Ah yes, my bad. I was trying to do something else and forgot to remove this.

chahak13 commented on lib/matplotlib/widgets.py:
```diff
@@ -1086,7 +1098,9 @@ def get_status(self):
         Return a tuple of the status (True/False) of all of the check buttons.
         """
-        return [l1.get_visible() for (l1, l2) in self.lines]
+        return [False if colors.same_color(
+            self._crosses.get_facecolors()[i], colors.to_rgba("none"))
+                else True for i in range(len(self.labels))]
```
This is what I tried first too but wouldn't work since `same_color` returns a single boolean and not element-wise comparison.

chahak13 replied:
It needs both the arguments to be of the same shape and then just returns `(c1 ==  c2).all()` so can't compare a list of colors with `"none"`

QuLogic replied:
Ah oops, that's unfortunate.

This can still be simplified though:
```suggestion
        return [not colors.same_color(color, colors.to_rgba("none"))
                for color in self._crosses.get_facecolors()]
```

chahak13 replied:
Fair enough. That's better. Thanks, done!

chahak13 commented on lib/matplotlib/widgets.py:
```diff
@@ -1002,43 +1002,25 @@ def __init__(self, ax, labels, actives=None):
-            ax.add_line(l2)
+        self._squares = ax.scatter(
+            [0.15] * len(ys), ys, marker='s', s=text_size**2,
+            c=["none" for _ in range(len(ys))],
```
The reason I used an array instead of just `c="none"` is because when I create the `._rectangles` attribute, `facecolor=self._squares.get_facecolor()` in `Rectangle gives a `ValueError` because `none` is returning an empty array. On the other hand, using `["none"] * len(ys)` returns a zero array as expected. This was unexpected, I'm trying to check why this is happening.

chahak13 replied:
The `to_rgba_array` returns a `(0, 4)` shape array for `c="none"` which then gets mapped to the facecolors. This can't be used as a value to `facecolors` though since it requires a non-zero shaped array.

chahak13 replied:
Though, since I'm using this information only in the Rectangle patch which will be removed later on, I can work around this and make the scatter function for the squares cleaner with `c="none"`. I'll do that.

tacaswell commented:
Running this example: https://github.com/matplotlib/matplotlib/blob/main/examples/widgets/check_buttons.py

The  `CheckButtons` work and show the initial state correctly, but the x 's do not come back when you go from off -> on.

tacaswell commented:
I'm going to push this to 3.8, but if it is indeed ready by the end of the year I am 👍🏻 on it getting merged and being re-milestoned to 3.7.

chahak13 commented:
That issue was not supposed to be there anymore....taking a look right away. Sorry this is taking too long, I keep missing small things with this one for some reason.

chahak13 commented:
@tacaswell corrected that. Everything seems to be working locally. Can you check once? Thanks!

tacaswell commented:
@chahak13 Could you squash this to a smaller number of commits?

chahak13 commented:
That should do it. I've kept the major, important commits and squashed all changes into them.

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -1072,9 +1068,20 @@ def set_active(self, index):
+        cross_facecolors[index] = colors.to_rgba(
+            "black"
+            if colors.same_color(
+                    cross_facecolors[index], colors.to_rgba("none")
```
Extra indent

chahak13 replied:
That was a weird editor quirk. Removed, thanks.

QuLogic commented on doc/api/next_api_changes/deprecations/24474_CM.rst:
```diff
@@ -0,0 +1,3 @@
+``CheckButtons.rectangles`` and ``CheckButtons.lines``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+``CheckButtons.rectangles`` and ``CheckButtons.lines`` are deprecated.  (``CheckButtons`` now draws itself using `~.Axes.scatter`.)
```
Wrap at 80 characters also.

chahak13 replied:
Ah okay. Missed that.

QuLogic reviewed (approved):
I think I'm okay with this now, but you may want to squash it down again?

chahak13 commented:
@QuLogic done. Should be good.
